### PR TITLE
supports `#source=true&workspace=name`

### PR DESCRIPTION
**Pull request by Larry1123**
Supports the same workspace logic provided in `plugin-git`

**What's the problem this PR addresses?**

This adds extra logic to tarball downloads to be able to support source repos and workspace names within.
This is intended to support other plugins to fetching via this this new functionally along as directly via end-users.

**How did you fix it?**

If `#source` is provided as part of the url it will unpack the tar as if it was a repo. It will call `scriptUtils.prepareExternalProject` on the extracted project with a provided `#workspace=name` to pack the project.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Comment by Larry1123** [comment]
I plan on updating docks for end users this new support is wanted.

File: packages/plugin-http/sources/TarballHttpFetcher.ts [hunk 1]
```diff
@@ -38,14 +40,50 @@ export class TarballHttpFetcher implements Fetcher {
...
+    let url = locator.reference;
+    const hashIndex = url.indexOf(`#`);
+    if (hashIndex !== -1) {
+      url = url.slice(0, hashIndex);
+      const extra = querystring.parse(url.slice(hashIndex + 1));
+      source = !!extra?.source;
+      if (source && extra.workspace) {
+        workspace = Array.isArray(extra.workspace) ? extra.workspace[0] : extra.workspace;
+      }
+    }
```
Comment (arcanis) [inline packages/plugin-http/sources/TarballHttpFetcher.ts:54]: Needs to use `structUtils.parseRange`

**Review by arcanis — COMMENTED** [review]
I'm still a bit unsure of the exact use case, but the implementation seems reasonable enough. If you can detail more how you envision it being used (for plugins, if I understand correctly?) it would probably be useful for later reference.

**Comment by Larry1123** [comment]
https://github.com/Larry1123/berry/blob/PoC/tarballRefactor/packages/plugin-github/sources/githubUtils.ts#L92
https://github.com/Larry1123/berry/blob/PoC/tarballRefactor/packages/plugin-github/sources/GithubFetcher.ts#L18
This is an example of `plugin-github` being able to use this internally to. Similarly a `plugin-gitlab` would be able to make use of this also.
End-users could make use of it directly also for any source that can be provided as a tarball. A plugin would not necessarily be needed if the end user wants to make use of this.

**Pull request closed without merging**